web/src/message_view.ts

```diff
@@ -913,8 +949,37 @@ function load_local_messages(msg_data: MessageListData, superset_data: MessageLi
 
     const in_msgs = superset_data.all_messages();
     msg_data.add_messages(in_msgs);
+    if (msg_data.visibly_empty()) {
+        return false;
+    }
+
+    // We cannot rely on fetch status of data for which
+    // `zerver.lib.narrow.ok_to_include_history` is not `True` when
+    // populating narrow for which it is `True` since older
+    // messages might be filtered due to them not having `UserMessage`
+    // entries. See `zerver.lib.narrow.fetch_messages`.
+    let ok_to_copy_status = true;
+    if (!superset_data.filter.includes_history() && msg_data.filter.includes_history()) {
+        ok_to_copy_status = false;
+    }
+
+    if (!ok_to_copy_status && msg_data.filter.has_operator("channel")) {
+        // We can still rely on fetch status if the superset_data
+        // has the first message of the channel.
+        const channel_id_str = msg_data.filter.operands("channel")[0];
+        assert(channel_id_str !== undefined);
+        const channel_id = Number.parseInt(channel_id_str, 10);
+        const oldest_possible_msg_id = sub_store.get(channel_id)?.first_message_id;
+        if (oldest_possible_msg_id && superset_data.get(oldest_possible_msg_id)) {
+            ok_to_copy_status = true;
+        }
+    }
```

Comment:
@timabbott I think we can merge this PR by removing this optimization. 

Our blocker was this:

> Oh, but I guess this is still not entirely correct, in that we don't handle the case that the current user was not continuously subscribed to the channel -- so the all_messages_data cache might be incomplete for that reason.

The `if` condition above this will prevent populating the `msg_data` from `all_messages_data` in that case.

Reply:
Past czo discussion - https://chat.zulip.org/#narrow/channel/3-backend/topic/Handling.20users.20being.20unsubscribed.20to.20a.20channel.20for.20a.20while

Reply:
Let me test this and see if it fixes the bug.

Reply:
Yeah, that fixes the issue.

Reply:
Sorry, I didn't test completely, it would still won't show messages which were send while user was unsubscribed if a message was sent after user subscribed.